ros2cli/ros2cli/daemon/daemonize.py

```diff
@@ -0,0 +1,113 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import copyreg
+import os
+import pickle
+import platform
+import socket
+import subprocess
+import sys
+import threading
+
+from ros2cli.helpers import wait_for
+
+
+class PicklerForProcess(pickle.Pickler):
+
+    def __init__(self, process, *args, **kwargs):
+        super().__init__(process.stdin, *args, **kwargs)
+        self.process = process
+        self.dispatch_table = copyreg.dispatch_table.copy()
+        self.dispatch_table[socket.socket] = self.reduce_socket
+        self.dispatch_table[threading.Event] = self.reduce_event
+
+    def reduce_event(self, obj):
+        return threading.Event, ()
+
+    @staticmethod
+    def load_socket(data):
+        if platform.system() == 'Windows':
+            return socket.fromshare(data)
+        return socket.socket(fileno=data)
+
+    def reduce_socket(self, obj):
+        if platform.system() == 'Windows':
+            data = obj.share(self.process.pid)
+        else:
+            data = obj.fileno()
+        return PicklerForProcess.load_socket, (data,)
+
+    def dump(self, *args, **kwargs):
+        super().dump(*args, **kwargs)
+        self.process.stdin.flush()
+
+
+def main():
+    func = pickle.load(sys.stdin.buffer)
+    sys.stdin.close()
+    os.close(0)  # force C stream close
+    return func()
+
+
+def daemonize(func, tags={}, timeout=None, debug=False):
+    prog = f'from {__name__} import main; main()'
+    cmd = [sys.executable, '-c', prog]
+    for name, value in tags.items():
+        flag = '--' + name.replace('_', '-')
+        cmd += [flag, str(value)]
+    kwargs = {}
+    if platform.system() == 'Windows':
+        if not debug:
+            kwargs.update(creationflags=subprocess.DETACHED_PROCESS)
+        # avoid showing cmd windows for subprocess
+        si = subprocess.STARTUPINFO()
+        si.dwFlags = subprocess.STARTF_USESHOWWINDOW
+        si.wShowWindow = subprocess.SW_HIDE
+        kwargs['startupinfo'] = si
+        # don't keep handle of current working directory in daemon process
+        kwargs.update(cwd=os.environ.get('SYSTEMROOT', None))
+
+    kwargs['stdin'] = subprocess.PIPE
+    if not debug:
+        kwargs['stdout'] = subprocess.DEVNULL
+        kwargs['stderr'] = subprocess.DEVNULL
+    kwargs['close_fds'] = False
+
+    process = subprocess.Popen(cmd, **kwargs)
+
+    pickler = PicklerForProcess(process)
+
+    pickler.dump(func)
+
+    if timeout is not None:
+        def daemon_ready():
+            try:
+                pickler.dump(None)
+                return False
+            except OSError:
+                return True
```

Comment:
how does this work?

Reply:
Also a good question. 

A daemon that is ready is a daemon that has deserialized the callable sent. When passing a socket, like this patch does when binding the server instance, that means it has been successfully reconstructed on the other end. By waiting on the daemon to be ready, we make sure it is safe to destroy the server and release the socket in the parent process. But only the child process knows when it's ready, so we need a form of communication. To not interfere with `stdout` and `stderr` redirection nor create more IPCs, I chose to use the act of closing the `stdin` pipe as the signal.

So, to summarize, the time flow of `daemonize()` goes like this: 

1. Parent process spawns child process, piping `stdin`.
2. Parent process pickles callable and sends it through `stdin`. 
4. Child process unpickles callable from `stdin`.
5. Child process closes `stdin`.
6. Parent process pickles `None` and sends it through `stdin`.
7. Parent process gets an `stdin` pipe closed exception.
8. Parent process checks child process is still alive.

Reply:
I see, that sounds like a fair way of solving the issue.
I would also leave a minimal comment here, the logic doesn't seem to be immediately obvious.

Reply:
Comments and documentation added in https://github.com/ros2/ros2cli/pull/652/commits/4ceb6fef7d400f7f3db94dfdd6c9f71f929fcab4.